### PR TITLE
MRG Ica plot

### PR DESCRIPTION
**Pull request by jaeilepp**
Here's a plotter for ICA components (item at https://github.com/mne-tools/mne-python/issues/2213). The interactive selection doesn't work yet and neither does scrolling horizontally, but I wanted to show what I've done so far, because I'm not sure if this is what you guys wanted in the first place. I'm no expert in using ICA, so comments are welcome.
screenshot from 2015-06-30 15 33 23
The plotter has basically the same interactive functions as does other browser functions. This PR also includes quite a bit of re-factoring of the existing browser functions.

### CODE REVIEW

**Comment by Eric89GXL** [comment]
I was going to suggest that you do some heavy refactoring to unify bits of
code but it sounds like you're already on it! Excellent.

File: mne/preprocessing/ica.py [hunk 1]
```diff
@@ -1572,6 +1572,39 @@ def _check_n_pca_components(self, _n_pca_comp, verbose=None):
 
         return _n_pca_comp
 
+    def plot_raw_components(self, raw, bads=[], title=None, duration=10.0,
```
Comment (dengemann) [inline mne/preprocessing/ica.py]: we should avoid mutable defaults e.g. lists.

**Comment by agramfort** [comment]
cool

let us know when it's ready. If not yet.

**Comment by mainakjas** [comment]
awesome ! how should I test this? Is there an example I can run?

**Comment by jaeilepp** [comment]
I added a line in the example that uses this function. If you're gonna run it many times, it's probably wise to save the ica solution and use that.

**Comment by jaeilepp** [comment]
I made it so that the plotting function now returns the updated list of components marked for exclusion, as selected by the user. Would it be okay to exclude the components straight away on closing the figure (ica.exclude)? At the moment it only updates the list, but is inconsistent with other plot functions because it returns a tuple of figure and list instead of just the figure.

File: mne/preprocessing/ica.py [hunk 2]
```diff
@@ -1572,6 +1572,53 @@ def _check_n_pca_components(self, _n_pca_comp, verbose=None):
...
+            (click on line). May not work on all systems / platforms.
+            Defaults to False.
+
+        Returns
+        -------
+        fig : Instance of matplotlib.figure.Figure
+            The figure.
+        exclude : list
+            Updated list of components marked for exclusion.
+        """
```
Comment (dengemann) [inline mne/preprocessing/ica.py]: did we already discuss why not simply updating ICA.exclude?

**Comment by dengemann** [comment]
> Would it be okay to exclude the components straight away on closing the figure (ica.exclude)? 

I would opt for that.

**Comment by agramfort** [comment]
+1

File: mne/preprocessing/ica.py (hunk 2 above)
Comment (agramfort) [inline mne/preprocessing/ica.py]: No

File: examples/preprocessing/plot_ica_from_raw.py [hunk 3]
```diff
@@ -62,6 +62,7 @@
 
 show_picks = np.abs(scores).argsort()[::-1][:5]
 
+ica.plot_raw_components(raw, exclude=ecg_inds, title=title % 'ecg')
```
Comment (mainakjas) [inline examples/preprocessing/plot_ica_from_raw.py]: This example is quite crazy. It opens 12 or 13 figures ! anything can be done about it? cc @dengemann

**Comment by mainakjas** [comment]
There is no help button here. I suppose you'll add that soon?

**Comment by jaeilepp** [comment]
The help button positioning is partly dependent on the layout remake at https://github.com/mne-tools/mne-python/pull/2214. I'd rather wait for it to be merged before implementing it here.

File: examples/preprocessing/plot_ica_from_raw.py (hunk 3 above)
Comment (dengemann) [inline examples/preprocessing/plot_ica_from_raw.py]: I would like to leave it like this, it has been very useful for tracking bugs in the past.

**Comment by dengemann** [comment]
@jaeilepp @mainakjas @agramfort I still think we should just overwrite plot_sources or so. Not sure we should add new methods.

**Comment by mainakjas** [comment]
On Fri, Jul 3, 2015 at 10:18 AM, Denis A. Engemann <notifications@github.com
> wrote:

> @jaeilepp <https://github.com/jaeilepp> @mainakjas
> <https://github.com/mainakjas> @agramfort <https://github.com/agramfort>
> I still think we should just overwrite plot_sources or so. Not sure we
> should add new methods.
>
+1 I agree. What we have is a more sophisticated version of plot_sources,
right?

**Comment by dengemann** [comment]
> plot_sources,
right?

yes and plot_sources was the weakest of the plotting functions at least for raw. The evoked counterpart is really useful.

**Comment by dengemann** [comment]
... which reminds us of the fact that we should also recycle the epochs viz here. I would suspect that is at least as useful as the raw viz for ICA.

**Comment by agramfort** [comment]
+1 for recycling raw and epochs browser here.

**Comment by jaeilepp** [comment]
Epoch plotter implemented. Is there any sense in marking bad epochs here? Should it be disabled? Also, is the exclude parameter needed? Ica object already has an exclude attribute, which is basically the same. Same goes for raw plotter. So basically the use case would change so that the user updates the ica.exclude before plotting to see the components marked for exclusion.

**Comment by dengemann** [comment]
No sense, so yes, it should be disabled

**Comment by jaeilepp** [comment]
I changed the plot_sources to use these new functions for epochs and raw instances. See if you're happy.

File: doc/source/whats_new.rst [hunk 4]
```diff
@@ -14,6 +14,8 @@ Changelog
 
     - Speed up TF-MxNE inverse solver with block coordinate descent by `Daniel Strohmeier`_ and `Yousra Bekhti`_
 
+	- Add ICA plotters for raw and epoch components by `Jaakko Leppakangas`_
```
Comment (agramfort) [inline doc/source/whats_new.rst]: bad indent

**Comment by agramfort** [comment]
works great for me

did not review full code though.

Usability feedback. Not for this PR but it would be great to preview the
topography of source from the browser. I see some nice alpha waves on raw
examples and it's frustrating not to be able to see the associated
topography.

**Comment by jaeilepp** [comment]
> Usability feedback. Not for this PR but it would be great to preview the topography of source from the browser.

Did you mean like this? I implemented it here, since it was quite easy to do.
screenshot from 2015-07-07 12 23 25
The component can be shown by clicking on the y axis label. I don't think it's going to work on every data since it relies on getting the layout from the data.
The indexing is actually inconsistent (ICA 109 = IC 108). I'll change that...

**Comment by dengemann** [comment]
> The indexing is actually inconsistent (ICA 109 = IC 108). I'll change that...

Now it's a good moment to attack that (cc @teonlamont). I'd do it at the plotting level, not at the level of the sources as raw function.

I'm still wondering though what's better. It's actually correct: ICA component #109 has the python index 108, wheras component #0 does not exist. We do the same for the channels btw.

**Comment by jaeilepp** [comment]
> I'm still wondering though what's better. It's actually correct: ICA component #109 has the python index 108, wheras component #0 does not exist. We do the same for the channels btw.

Yeah, I intuitively thought that the indexing should start from 1, but all the existing plotting functions seem to be using 0 for ICA.

**Comment by jaeilepp** [comment]
I guess this is ready for review, but it still needs some testing.

**Comment by teonlamont** [comment]
closes #2000 #2043.
are we able to mark bad ICs from the browser like in the raw/epochs viewer?

also, what's the verdict on the indexing, are we plotting the ICs to start at 1 e.g. `ICA 001` for all the plots now? do we still exclude based on the python index? if so, should we support exclude to take IC label to be in line with `info['bads']`? I think @agramfort and me mentioned it as a solution at one point https://github.com/mne-tools/mne-python/pull/2019#issuecomment-96275197. for me, as long as we have a consistent use of names across the board, I'm happy.

btw, the topo view is super useful

**Comment by agramfort** [comment]
you need to rebase

**Comment by agramfort** [comment]
are you showing only the topo for the magnetometers? what if you have fitted ica with grads too and even eeg?

**Comment by jaeilepp** [comment]
It's using the defaults of plot_components. Documentation says:
>ch_type : 'mag' | 'grad' | 'planar1' | 'planar2' | 'eeg' | None
            The channel type to plot. For 'grad', the gradiometers are
            collected in pairs and the RMS for each pair is plotted.
            If None, then channels are chosen in the order given above.

**Comment by agramfort** [comment]
ok. Could we show all ch types available? grad, mag and EEG? I am curious
to visualize to the topos of grad vs eeg for example.

**Comment by jaeilepp** [comment]
You mean a field just showing the channel types on plot_components or an interactive selection of channels?

**Comment by agramfort** [comment]
I mean showing all topos rather than just one when you click on channel
name.

**Comment by jaeilepp** [comment]
Okay, I think this is ready for review. There might be some glitches with it, since a lot have changed with the browsers functions. At least I can't find any major flaws, but I'm not an expert with ICA.
screenshot from 2015-07-10 13 32 20

**Comment by teonlamont** [comment]
channel numbers should be formatted `%03d`.

**Comment by teonlamont** [comment]
So how do you remove bad ICs and how do you plot the topography? I think I read that topo come from clicking the ICA label? the only problem i see with that for raw.plot, epochs.plot, that label clicking is for marking the bad channels.

**Comment by agramfort** [comment]
@teonlamont you can click on the lines. I think it's ok.

any other opinion?

**Comment by teonlamont** [comment]
just the number format for the channel names

**Comment by jaeilepp** [comment]
> any other opinion?

At the moment clicking the data does the 'primary' function (bad channel for raw, bad epoch for epoch, bad component for ICA). The 'secondary' function is done by clicking the label (bad channel for raw and epoch, topography for ICA). This seems the most intuitive solution for me.
Ready for review.

**Comment by Eric89GXL** [comment]
Don't worry @choldgraf there are plenty of other things to do. You could
potentially improve the data viz browsers like you've wanted to for a
while?

**Comment by Eric89GXL** [comment]
Whoops wrong thread!

File: mne/viz/ica.py [hunk 5]
```diff
@@ -510,3 +524,196 @@ def _plot_ica_overlay_evoked(evoked, evoked_cln, title, show):
...
+    if stop is None:
+        stop = start + 20
+        stop = min(stop, raw.times[-1])
+    duration = stop - start
+    if duration <= 0:
+        raise RuntimeError('Stop must be larger than start.')
+    t_end = int(duration * raw.info['sfreq'])
+    times = raw.times[0:t_end]
+    bad_color = (1., 0., 0.)
+    inds = range(len(picks))
```
Comment (dengemann) [inline mne/viz/ica.py]: you should wrap in in list for py3k versions

File: mne/viz/ica.py [hunk 6]
```diff
@@ -510,3 +524,196 @@ def _plot_ica_overlay_evoked(evoked, evoked_cln, title, show):
...
+    data = ica._transform_epochs(epochs, concatenate=True)
+    c_names = ['ICA %03d' % x for x in range(ica.n_components_)]
+    scalings = {'misc': 5.0}
+    info = create_info(ch_names=c_names, sfreq=epochs.info['sfreq'])
+    info['projs'] = list()
+    info['bads'] = [c_names[x] for x in exclude]
+    if title is None:
+        title = 'ICA components'
+    if picks is None:
+        picks = range(len(c_names))
```
Comment (dengemann) [inline mne/viz/ica.py]: outer list

File: mne/viz/raw.py [hunk 7]
```diff
@@ -790,3 +503,163 @@ def plot_raw_psd(raw, tmin=0., tmax=np.inf, fmin=0, fmax=np.inf, proj=False,
...
+
+def _plot_raw_traces(params, inds, color, bad_color, event_lines=None,
+                     event_color=None):
+    """Helper for plotting raw"""
+    lines = params['lines']
+    info = params['info']
+    n_channels = params['n_channels']
+    params['bad_color'] = bad_color
+    # do the plotting
+    tick_list = []
```
Comment (dengemann) [inline mne/viz/raw.py]: list()

**Comment by dengemann** [comment]
A few observations:
- in the standard example the ECG seems to be invisible on the time course, the topo looks right.
- the plot is not efficiently used, many white spaces
- the original ECG / channels are not shown, they may be very helpful for detecting visually correlations. use ``ica.get_sources`` for this.

I found a few minor issues, I'll open a PR to your branch.

**Comment by dengemann** [comment]
Btw. for epochs the timeseries looks correct.

**Comment by dengemann** [comment]
@jaeilepp can you rebase?

**Comment by dengemann** [comment]
@jaeilepp something went wrong with your rebasing. I see commits listed multiple times :(

File: mne/viz/ica.py (hunk 5 above)
Comment (agramfort) [inline mne/viz/ica.py]: +1

File: mne/viz/ica.py (hunk 6 above)
Comment (agramfort) [inline mne/viz/ica.py]: or maybe use a range fix from six

is this all covered by tests?

**Comment by agramfort** [comment]
besides LGTM. Works great on my box.

@jaeilepp you'll need to rebase.

**Comment by dengemann** [comment]
@agramfort do your IC timeseries in the raw example look fine?

**Comment by jaeilepp** [comment]
Ready for review.

**Comment by agramfort** [comment]
I am happy

+1 for merge

@dengemann ?

**Comment by dengemann** [comment]
@jaeilepp we still don't see the EOG channels / ECG channels in the ICA time series, do we? I'll merge but please open an issue for later. Thanks for the wonderful work!

**Comment by agramfort** [comment]
great !

**Pull request merged**